{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d79bdabc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘arrow’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(optparse)\n",
    "library(arrow)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(yaml)\n",
    "\n",
    "source(\"watershed.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4b809d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trace(what = \"print\", where = getNamespace(\"base\"), exit = flush.console, print = FALSE)\n",
    "# trace(what = \"cat\", where = getNamespace(\"base\"), exit = flush.console, print = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d6a52bd2",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "library(methods)\n",
    "Snakemake <- setClass(\n",
    "    \"Snakemake\",\n",
    "    slots = c(\n",
    "        input = \"list\",\n",
    "        output = \"list\",\n",
    "        params = \"list\",\n",
    "        wildcards = \"list\",\n",
    "        threads = \"numeric\",\n",
    "        log = \"list\",\n",
    "        resources = \"list\",\n",
    "        config = \"list\",\n",
    "        rule = \"character\",\n",
    "        bench_iteration = \"numeric\",\n",
    "        scriptdir = \"character\",\n",
    "        source = \"function\"\n",
    "    )\n",
    ")\n",
    "snakemake <- Snakemake(\n",
    "    input = list('/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done', \"featureset_config\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml', \"data_pq\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet', \"data_pq_done\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done', \"subset_pq\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet', \"subset_pq_done\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done'),\n",
    "    output = list('/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done', \"train_data_dir\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet', \"test_data_dir\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet', \"model_joblib\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib', \"features_yaml\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml', \"touch_file\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done'),\n",
    "    params = list('/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets', '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models', 'train_simplecv.py', \"output_basedir\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm', \"feature_dir\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets', \"model_dir\" = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models', \"nb_script\" = 'train_simplecv.py'),\n",
    "    wildcards = list('gtex_v8_old_dna', 'dna_only', 'watershed', 'lightgbm', \"ds_dir\" = 'gtex_v8_old_dna', \"subset\" = 'dna_only', \"feature_set\" = 'watershed', \"model_type\" = 'lightgbm'),\n",
    "    threads = 1,\n",
    "    log = list(),\n",
    "    resources = list('tmpdir', 'ntasks', 'mem_mb', 'mem_mib', \"tmpdir\" = '/scratch/tmp/hoelzlwi', \"ntasks\" = 1, \"mem_mb\" = 120000, \"mem_mib\" = 114441),\n",
    "    config = list(\"projectTitle\" = 'REP', \"n_splits\" = 6, \"dirs\" = list(\"CACHE_DIR\" = '/s/project/rep/cache', \"RAW_DATA_DIR\" = '/s/project/rep/raw', \"PROCESSED_DATA_DIR\" = '/s/project/rep/processed', \"MODEL_DIR\" = '/s/project/rep/processed/training_results_v14'), \"canonical_transcript_pq\" = '/s/project/rep/processed/training_results_v14/general/gtex_canonical_transcript.parquet', \"expressed_genes_pq\" = '/s/project/rep/processed/training_results_v14/gtex_v8/gene_is_expressed_per_subtissue.parquet', \"system\" = list(\"MANE_select_url\" = 'https://ftp.ncbi.nlm.nih.gov/refseq/MANE/MANE_human/release_1.2/MANE.GRCh38.v1.2.ensembl_genomic.gtf.gz', \"gencode_basepath\" = '/s/genomes/Gencode/Gencode_human/release_{gencode_version}', \"vep\" = list(\"version\" = 108, \"vep_bin\" = 'vep', \"perl_bin\" = 'perl', \"vep_cache_dir\" = '/s/raw/vep/{vep_version}', \"cadd_dir\" = '/s/raw/cadd/v1.6/{human_genome_assembly}', \"loftee_data_dir\" = '/s/raw/loftee/{human_genome_assembly}', \"loftee_src_path\" = '/s/raw/loftee/{human_genome_assembly}_src'), \"absplice\" = list(\"cache\" = c('/s/project/rep/processed/absplice_cache/{human_genome_version}/veff.snp.parquet')))),\n",
    "    rule = 'models__train_simplecv',\n",
    "    bench_iteration = as.numeric(NA),\n",
    "    scriptdir = '/data/nasif12/home_if12/hoelzlwi/Projects/REP/rep_scripts/smk/ds_dir/models',\n",
    "    source = function(...){\n",
    "        wd <- getwd()\n",
    "        setwd(snakemake@scriptdir)\n",
    "        source(...)\n",
    "        setwd(wd)\n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "######## snakemake preamble end #########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02ca7827",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>[[1]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml'</dd>\n",
       "\t<dt>[[2]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet'</dd>\n",
       "\t<dt>[[3]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done'</dd>\n",
       "\t<dt>[[4]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet'</dd>\n",
       "\t<dt>[[5]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done'</dd>\n",
       "\t<dt>$featureset_config</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml'</dd>\n",
       "\t<dt>$data_pq</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet'</dd>\n",
       "\t<dt>$data_pq_done</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done'</dd>\n",
       "\t<dt>$subset_pq</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet'</dd>\n",
       "\t<dt>$subset_pq_done</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[{[[1]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/config.yaml'\n",
       "\\item[{[[2]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/data.parquet'\n",
       "\\item[{[[3]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/data.parquet.done'\n",
       "\\item[{[[4]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/subset/dna\\_only.parquet'\n",
       "\\item[{[[5]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/subset/dna\\_only.parquet.done'\n",
       "\\item[\\$featureset\\_config] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/config.yaml'\n",
       "\\item[\\$data\\_pq] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/data.parquet'\n",
       "\\item[\\$data\\_pq\\_done] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/fset@watershed/data.parquet.done'\n",
       "\\item[\\$subset\\_pq] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/subset/dna\\_only.parquet'\n",
       "\\item[\\$subset\\_pq\\_done] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/subset/dna\\_only.parquet.done'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "[[1]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml'\n",
       "[[2]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet'\n",
       "[[3]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done'\n",
       "[[4]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet'\n",
       "[[5]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done'\n",
       "$featureset_config\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml'\n",
       "$data_pq\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet'\n",
       "$data_pq_done\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done'\n",
       "$subset_pq\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet'\n",
       "$subset_pq_done\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml\"\n",
       "\n",
       "[[2]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet\"\n",
       "\n",
       "[[3]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done\"\n",
       "\n",
       "[[4]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet\"\n",
       "\n",
       "[[5]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done\"\n",
       "\n",
       "$featureset_config\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/config.yaml\"\n",
       "\n",
       "$data_pq\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet\"\n",
       "\n",
       "$data_pq_done\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/fset@watershed/data.parquet.done\"\n",
       "\n",
       "$subset_pq\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet\"\n",
       "\n",
       "$subset_pq_done\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/subset/dna_only.parquet.done\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snakemake@input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64cd0de9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>[[1]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet'</dd>\n",
       "\t<dt>[[2]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet'</dd>\n",
       "\t<dt>[[3]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib'</dd>\n",
       "\t<dt>[[4]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml'</dd>\n",
       "\t<dt>[[5]]</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done'</dd>\n",
       "\t<dt>$train_data_dir</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet'</dd>\n",
       "\t<dt>$test_data_dir</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet'</dd>\n",
       "\t<dt>$model_joblib</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib'</dd>\n",
       "\t<dt>$features_yaml</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml'</dd>\n",
       "\t<dt>$touch_file</dt>\n",
       "\t\t<dd>'/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[{[[1]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/train.parquet'\n",
       "\\item[{[[2]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/test.parquet'\n",
       "\\item[{[[3]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/model.joblib'\n",
       "\\item[{[[4]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/features.yaml'\n",
       "\\item[{[[5]]}] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/done'\n",
       "\\item[\\$train\\_data\\_dir] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/train.parquet'\n",
       "\\item[\\$test\\_data\\_dir] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/test.parquet'\n",
       "\\item[\\$model\\_joblib] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/model.joblib'\n",
       "\\item[\\$features\\_yaml] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/features.yaml'\n",
       "\\item[\\$touch\\_file] '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/models/dna\\_only/watershed@train\\_simplecv.py\\#lightgbm/done'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "[[1]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet'\n",
       "[[2]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet'\n",
       "[[3]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib'\n",
       "[[4]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml'\n",
       "[[5]]\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done'\n",
       "$train_data_dir\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet'\n",
       "$test_data_dir\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet'\n",
       "$model_joblib\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib'\n",
       "$features_yaml\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml'\n",
       "$touch_file\n",
       ":   '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet\"\n",
       "\n",
       "[[2]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet\"\n",
       "\n",
       "[[3]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib\"\n",
       "\n",
       "[[4]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml\"\n",
       "\n",
       "[[5]]\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done\"\n",
       "\n",
       "$train_data_dir\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/train.parquet\"\n",
       "\n",
       "$test_data_dir\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/test.parquet\"\n",
       "\n",
       "$model_joblib\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/model.joblib\"\n",
       "\n",
       "$features_yaml\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/features.yaml\"\n",
       "\n",
       "$touch_file\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/models/dna_only/watershed@train_simplecv.py#lightgbm/done\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snakemake@output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "38dd8ace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$features</dt>\n",
       "\t\t<dd><strong>$vep</strong> = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL'</dd>\n",
       "\t<dt>$pivot_tissues</dt>\n",
       "\t\t<dd>TRUE</dd>\n",
       "\t<dt>$snakemake</dt>\n",
       "\t\t<dd><strong>$input</strong> = <strong>$features</strong> = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL/data.parquet.done'</dd>\n",
       "\t<dt>$variables</dt>\n",
       "\t\t<dd><strong>$vep</strong> = <ol>\n",
       "\t<li><strong>$features</strong> = <style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'cadd_raw.max'</li><li>'LoF_HC.max'</li><li>'LoF_LC.max'</li><li>'5_prime_UTR_variant.max'</li><li>'3_prime_UTR_variant.max'</li><li>'TF_binding_site_variant.max'</li><li>'downstream_gene_variant.max'</li><li>'intergenic_variant.max'</li><li>'intron_variant.max'</li><li>'missense_variant.max'</li><li>'non_coding_transcript_exon_variant.max'</li><li>'regulatory_region_variant.max'</li><li>'splice_acceptor_variant.max'</li><li>'splice_donor_variant.max'</li><li>'splice_region_variant.max'</li><li>'stop_gained.max'</li><li>'synonymous_variant.max'</li><li>'upstream_gene_variant.max'</li><li>'sift_score.pval_max_significant'</li><li>'polyphen_score.max'</li></ol>\n",
       "</li>\n",
       "</ol>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$features] \\textbf{\\$vep} = '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/feature\\_sets/aggregate\\_vep\\_spark.py@gt=any\\_tl=CANONICAL'\n",
       "\\item[\\$pivot\\_tissues] TRUE\n",
       "\\item[\\$snakemake] \\textbf{\\$input} = \\textbf{\\$features} = '/s/project/rep/processed/training\\_results\\_v14/gtex\\_v8\\_old\\_dna/feature\\_sets/aggregate\\_vep\\_spark.py@gt=any\\_tl=CANONICAL/data.parquet.done'\n",
       "\\item[\\$variables] \\textbf{\\$vep} = \\begin{enumerate}\n",
       "\\item \\textbf{\\$features} = \\begin{enumerate*}\n",
       "\\item 'cadd\\_raw.max'\n",
       "\\item 'LoF\\_HC.max'\n",
       "\\item 'LoF\\_LC.max'\n",
       "\\item '5\\_prime\\_UTR\\_variant.max'\n",
       "\\item '3\\_prime\\_UTR\\_variant.max'\n",
       "\\item 'TF\\_binding\\_site\\_variant.max'\n",
       "\\item 'downstream\\_gene\\_variant.max'\n",
       "\\item 'intergenic\\_variant.max'\n",
       "\\item 'intron\\_variant.max'\n",
       "\\item 'missense\\_variant.max'\n",
       "\\item 'non\\_coding\\_transcript\\_exon\\_variant.max'\n",
       "\\item 'regulatory\\_region\\_variant.max'\n",
       "\\item 'splice\\_acceptor\\_variant.max'\n",
       "\\item 'splice\\_donor\\_variant.max'\n",
       "\\item 'splice\\_region\\_variant.max'\n",
       "\\item 'stop\\_gained.max'\n",
       "\\item 'synonymous\\_variant.max'\n",
       "\\item 'upstream\\_gene\\_variant.max'\n",
       "\\item 'sift\\_score.pval\\_max\\_significant'\n",
       "\\item 'polyphen\\_score.max'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$features\n",
       ":   **$vep** = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL'\n",
       "$pivot_tissues\n",
       ":   TRUE\n",
       "$snakemake\n",
       ":   **$input** = **$features** = '/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL/data.parquet.done'\n",
       "$variables\n",
       ":   **$vep** = 1. **$features** = 1. 'cadd_raw.max'\n",
       "2. 'LoF_HC.max'\n",
       "3. 'LoF_LC.max'\n",
       "4. '5_prime_UTR_variant.max'\n",
       "5. '3_prime_UTR_variant.max'\n",
       "6. 'TF_binding_site_variant.max'\n",
       "7. 'downstream_gene_variant.max'\n",
       "8. 'intergenic_variant.max'\n",
       "9. 'intron_variant.max'\n",
       "10. 'missense_variant.max'\n",
       "11. 'non_coding_transcript_exon_variant.max'\n",
       "12. 'regulatory_region_variant.max'\n",
       "13. 'splice_acceptor_variant.max'\n",
       "14. 'splice_donor_variant.max'\n",
       "15. 'splice_region_variant.max'\n",
       "16. 'stop_gained.max'\n",
       "17. 'synonymous_variant.max'\n",
       "18. 'upstream_gene_variant.max'\n",
       "19. 'sift_score.pval_max_significant'\n",
       "20. 'polyphen_score.max'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$features\n",
       "$features$vep\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL\"\n",
       "\n",
       "\n",
       "$pivot_tissues\n",
       "[1] TRUE\n",
       "\n",
       "$snakemake\n",
       "$snakemake$input\n",
       "$snakemake$input$features\n",
       "[1] \"/s/project/rep/processed/training_results_v14/gtex_v8_old_dna/feature_sets/aggregate_vep_spark.py@gt=any_tl=CANONICAL/data.parquet.done\"\n",
       "\n",
       "\n",
       "\n",
       "$variables\n",
       "$variables$vep\n",
       "$variables$vep[[1]]\n",
       "$variables$vep[[1]]$features\n",
       " [1] \"cadd_raw.max\"                          \n",
       " [2] \"LoF_HC.max\"                            \n",
       " [3] \"LoF_LC.max\"                            \n",
       " [4] \"5_prime_UTR_variant.max\"               \n",
       " [5] \"3_prime_UTR_variant.max\"               \n",
       " [6] \"TF_binding_site_variant.max\"           \n",
       " [7] \"downstream_gene_variant.max\"           \n",
       " [8] \"intergenic_variant.max\"                \n",
       " [9] \"intron_variant.max\"                    \n",
       "[10] \"missense_variant.max\"                  \n",
       "[11] \"non_coding_transcript_exon_variant.max\"\n",
       "[12] \"regulatory_region_variant.max\"         \n",
       "[13] \"splice_acceptor_variant.max\"           \n",
       "[14] \"splice_donor_variant.max\"              \n",
       "[15] \"splice_region_variant.max\"             \n",
       "[16] \"stop_gained.max\"                       \n",
       "[17] \"synonymous_variant.max\"                \n",
       "[18] \"upstream_gene_variant.max\"             \n",
       "[19] \"sift_score.pval_max_significant\"       \n",
       "[20] \"polyphen_score.max\"                    \n",
       "\n",
       "\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "featureset_config = read_yaml(snakemake@input$featureset_config)\n",
    "featureset_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64c04675",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 9203102 × 215</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>individual</th><th scope=col>gene</th><th scope=col>zscore.Adipose.20..20Subcutaneous</th><th scope=col>zscore.Adipose.20..20Visceral.20.28Omentum.29</th><th scope=col>zscore.Adrenal.20Gland</th><th scope=col>zscore.Artery.20..20Aorta</th><th scope=col>zscore.Artery.20..20Coronary</th><th scope=col>zscore.Artery.20..20Tibial</th><th scope=col>zscore.Brain.20..20Amygdala</th><th scope=col>zscore.Brain.20..20Anterior.20cingulate.20cortex.20.28BA24.29</th><th scope=col>⋯</th><th scope=col>feature.vep.features.regulatory_region_variant.max</th><th scope=col>feature.vep.features.splice_acceptor_variant.max</th><th scope=col>feature.vep.features.splice_donor_variant.max</th><th scope=col>feature.vep.features.splice_region_variant.max</th><th scope=col>feature.vep.features.stop_gained.max</th><th scope=col>feature.vep.features.synonymous_variant.max</th><th scope=col>feature.vep.features.upstream_gene_variant.max</th><th scope=col>feature.vep.features.sift_score.pval_max_significant</th><th scope=col>feature.vep.features.polyphen_score.max</th><th scope=col>fold</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214113</td><td>-0.46848705</td><td>-0.24032781</td><td>-0.22958386</td><td> 0.69779491</td><td>NA</td><td>-1.33789635</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214114</td><td>         NA</td><td> 1.23622894</td><td> 0.02992738</td><td> 1.16820312</td><td>NA</td><td> 0.41661450</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214160</td><td> 0.45258430</td><td>-0.01396400</td><td> 2.95433736</td><td> 0.38630462</td><td>NA</td><td> 0.46225259</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214193</td><td>-2.44460797</td><td> 2.13922572</td><td> 0.16586153</td><td>-0.11648754</td><td>NA</td><td>-0.71341413</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214194</td><td>         NA</td><td>-1.79452336</td><td> 0.32472101</td><td>-2.26251125</td><td>NA</td><td>-0.01573116</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214253</td><td>-1.39064407</td><td>-1.29023123</td><td>-1.72721636</td><td>-2.89380598</td><td>NA</td><td>-0.40899506</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214274</td><td>-0.31787163</td><td> 1.09867418</td><td>         NA</td><td>-0.29028556</td><td>NA</td><td>-0.41056791</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214357</td><td>-0.34382582</td><td>-0.17527953</td><td>-0.11419190</td><td> 1.51654112</td><td>NA</td><td> 2.03312397</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214413</td><td>-0.81799686</td><td>-1.65288782</td><td>-1.32054901</td><td>-0.96304089</td><td>NA</td><td> 0.96759593</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214517</td><td>-0.35869706</td><td> 1.71566033</td><td> 1.27662230</td><td> 1.28102994</td><td>NA</td><td> 0.72199219</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214530</td><td>         NA</td><td>         NA</td><td>-1.90293229</td><td>-0.87361908</td><td>NA</td><td> 0.15419285</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214655</td><td> 0.56697595</td><td> 0.91750306</td><td>-0.79633600</td><td> 1.05349636</td><td>NA</td><td>-1.15545642</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214706</td><td> 0.57189840</td><td> 1.25597429</td><td>-0.54215997</td><td>-0.58148170</td><td>NA</td><td>-0.38550556</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214717</td><td>-1.73147583</td><td>-1.22039437</td><td>-0.68413210</td><td>-0.09883536</td><td>NA</td><td>-1.31531465</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214736</td><td>-0.85407978</td><td>-1.68542659</td><td>-0.05669600</td><td>-1.12853837</td><td>NA</td><td> 0.15348506</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214753</td><td> 1.03080797</td><td>-0.04829543</td><td>-0.13953565</td><td>-0.38410354</td><td>NA</td><td> 0.88414800</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214940</td><td> 0.04631884</td><td>-1.77462924</td><td> 0.76365566</td><td> 0.40670428</td><td>NA</td><td> 0.15609336</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214944</td><td>         NA</td><td> 0.18850540</td><td> 0.96971118</td><td> 0.28924718</td><td>NA</td><td> 0.80388254</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000214967</td><td>-0.31326768</td><td>-1.82583344</td><td> 0.76598984</td><td> 0.21049729</td><td>NA</td><td>-0.75453371</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215012</td><td> 0.64287394</td><td> 0.23868078</td><td> 0.21511005</td><td> 0.44112059</td><td>NA</td><td> 0.56495368</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215021</td><td> 0.30005670</td><td>-0.16562910</td><td> 0.53408575</td><td> 0.85071629</td><td>NA</td><td>-1.27914917</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215041</td><td>-0.99744004</td><td> 0.55672777</td><td> 0.77792305</td><td>-0.40734518</td><td>NA</td><td> 2.26693535</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215114</td><td> 0.99766344</td><td>-1.02371836</td><td> 0.99257654</td><td> 0.81619185</td><td>NA</td><td> 0.90462214</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215126</td><td> 0.84524620</td><td> 0.56395090</td><td>-0.37646827</td><td> 0.68983483</td><td>NA</td><td> 1.48095489</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215193</td><td>-0.06694359</td><td> 0.21530281</td><td>-0.52403498</td><td> 1.69092894</td><td>NA</td><td> 1.29820156</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215251</td><td>         NA</td><td> 0.20708491</td><td> 0.25430226</td><td> 0.07474785</td><td>NA</td><td>-0.63851118</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215252</td><td> 0.48236033</td><td>-1.26481926</td><td>-0.36134848</td><td> 0.14095147</td><td>NA</td><td> 0.71949059</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215301</td><td>-0.97372675</td><td>-0.47196299</td><td>-0.32129517</td><td> 1.17368984</td><td>NA</td><td> 0.05126289</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215305</td><td>-0.03369558</td><td>-1.12211978</td><td> 0.50003290</td><td>-0.44566694</td><td>NA</td><td>-0.66550738</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>GTEX-1211K</td><td>ENSG00000215440</td><td> 0.85139471</td><td> 0.65542102</td><td>-0.12812780</td><td> 0.53822845</td><td>NA</td><td> 0.16725840</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>NA</td><td>NA</td><td>3</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000116031</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000116035</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000116039</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000119457</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000122877</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000124935</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000125571</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000128203</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000130055</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000130475</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000131686</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000131969</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000132677</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000132975</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000133247</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000134765</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000137709</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000138111</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000142046</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000143520</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000145244</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000148291</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000151117</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000154928</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000155367</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000158050</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000158106</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000158220</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td>Inf</td><td>0.991</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000158786</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "\t<tr><td>GTEX-ZVP2</td><td>ENSG00000159455</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td>   NA</td><td> NA</td><td>   NA</td><td>test</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 9203102 × 215\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " individual & gene & zscore.Adipose.20..20Subcutaneous & zscore.Adipose.20..20Visceral.20.28Omentum.29 & zscore.Adrenal.20Gland & zscore.Artery.20..20Aorta & zscore.Artery.20..20Coronary & zscore.Artery.20..20Tibial & zscore.Brain.20..20Amygdala & zscore.Brain.20..20Anterior.20cingulate.20cortex.20.28BA24.29 & ⋯ & feature.vep.features.regulatory\\_region\\_variant.max & feature.vep.features.splice\\_acceptor\\_variant.max & feature.vep.features.splice\\_donor\\_variant.max & feature.vep.features.splice\\_region\\_variant.max & feature.vep.features.stop\\_gained.max & feature.vep.features.synonymous\\_variant.max & feature.vep.features.upstream\\_gene\\_variant.max & feature.vep.features.sift\\_score.pval\\_max\\_significant & feature.vep.features.polyphen\\_score.max & fold\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t GTEX-1211K & ENSG00000214113 & -0.46848705 & -0.24032781 & -0.22958386 &  0.69779491 & NA & -1.33789635 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214114 &          NA &  1.23622894 &  0.02992738 &  1.16820312 & NA &  0.41661450 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214160 &  0.45258430 & -0.01396400 &  2.95433736 &  0.38630462 & NA &  0.46225259 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214193 & -2.44460797 &  2.13922572 &  0.16586153 & -0.11648754 & NA & -0.71341413 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214194 &          NA & -1.79452336 &  0.32472101 & -2.26251125 & NA & -0.01573116 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214253 & -1.39064407 & -1.29023123 & -1.72721636 & -2.89380598 & NA & -0.40899506 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214274 & -0.31787163 &  1.09867418 &          NA & -0.29028556 & NA & -0.41056791 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214357 & -0.34382582 & -0.17527953 & -0.11419190 &  1.51654112 & NA &  2.03312397 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214413 & -0.81799686 & -1.65288782 & -1.32054901 & -0.96304089 & NA &  0.96759593 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214517 & -0.35869706 &  1.71566033 &  1.27662230 &  1.28102994 & NA &  0.72199219 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214530 &          NA &          NA & -1.90293229 & -0.87361908 & NA &  0.15419285 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214655 &  0.56697595 &  0.91750306 & -0.79633600 &  1.05349636 & NA & -1.15545642 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214706 &  0.57189840 &  1.25597429 & -0.54215997 & -0.58148170 & NA & -0.38550556 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214717 & -1.73147583 & -1.22039437 & -0.68413210 & -0.09883536 & NA & -1.31531465 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214736 & -0.85407978 & -1.68542659 & -0.05669600 & -1.12853837 & NA &  0.15348506 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214753 &  1.03080797 & -0.04829543 & -0.13953565 & -0.38410354 & NA &  0.88414800 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214940 &  0.04631884 & -1.77462924 &  0.76365566 &  0.40670428 & NA &  0.15609336 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214944 &          NA &  0.18850540 &  0.96971118 &  0.28924718 & NA &  0.80388254 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000214967 & -0.31326768 & -1.82583344 &  0.76598984 &  0.21049729 & NA & -0.75453371 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215012 &  0.64287394 &  0.23868078 &  0.21511005 &  0.44112059 & NA &  0.56495368 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215021 &  0.30005670 & -0.16562910 &  0.53408575 &  0.85071629 & NA & -1.27914917 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215041 & -0.99744004 &  0.55672777 &  0.77792305 & -0.40734518 & NA &  2.26693535 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215114 &  0.99766344 & -1.02371836 &  0.99257654 &  0.81619185 & NA &  0.90462214 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215126 &  0.84524620 &  0.56395090 & -0.37646827 &  0.68983483 & NA &  1.48095489 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215193 & -0.06694359 &  0.21530281 & -0.52403498 &  1.69092894 & NA &  1.29820156 & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215251 &          NA &  0.20708491 &  0.25430226 &  0.07474785 & NA & -0.63851118 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215252 &  0.48236033 & -1.26481926 & -0.36134848 &  0.14095147 & NA &  0.71949059 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215301 & -0.97372675 & -0.47196299 & -0.32129517 &  1.17368984 & NA &  0.05126289 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215305 & -0.03369558 & -1.12211978 &  0.50003290 & -0.44566694 & NA & -0.66550738 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t GTEX-1211K & ENSG00000215440 &  0.85139471 &  0.65542102 & -0.12812780 &  0.53822845 & NA &  0.16725840 & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA & NA & NA & 3\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000116031 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000116035 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000116039 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000119457 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000122877 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000124935 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000125571 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000128203 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000130055 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000130475 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000131686 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000131969 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000132677 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000132975 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000133247 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000134765 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000137709 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000138111 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000142046 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000143520 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000145244 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000148291 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000151117 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000154928 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000155367 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000158050 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000158106 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000158220 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE & Inf & 0.991 & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000158786 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\t GTEX-ZVP2 & ENSG00000159455 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ &    NA &    NA &    NA &    NA &    NA &    NA &    NA &  NA &    NA & test\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 9203102 × 215\n",
       "\n",
       "| individual &lt;chr&gt; | gene &lt;chr&gt; | zscore.Adipose.20..20Subcutaneous &lt;dbl&gt; | zscore.Adipose.20..20Visceral.20.28Omentum.29 &lt;dbl&gt; | zscore.Adrenal.20Gland &lt;dbl&gt; | zscore.Artery.20..20Aorta &lt;dbl&gt; | zscore.Artery.20..20Coronary &lt;dbl&gt; | zscore.Artery.20..20Tibial &lt;dbl&gt; | zscore.Brain.20..20Amygdala &lt;dbl&gt; | zscore.Brain.20..20Anterior.20cingulate.20cortex.20.28BA24.29 &lt;dbl&gt; | ⋯ ⋯ | feature.vep.features.regulatory_region_variant.max &lt;lgl&gt; | feature.vep.features.splice_acceptor_variant.max &lt;lgl&gt; | feature.vep.features.splice_donor_variant.max &lt;lgl&gt; | feature.vep.features.splice_region_variant.max &lt;lgl&gt; | feature.vep.features.stop_gained.max &lt;lgl&gt; | feature.vep.features.synonymous_variant.max &lt;lgl&gt; | feature.vep.features.upstream_gene_variant.max &lt;lgl&gt; | feature.vep.features.sift_score.pval_max_significant &lt;dbl&gt; | feature.vep.features.polyphen_score.max &lt;dbl&gt; | fold &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GTEX-1211K | ENSG00000214113 | -0.46848705 | -0.24032781 | -0.22958386 |  0.69779491 | NA | -1.33789635 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214114 |          NA |  1.23622894 |  0.02992738 |  1.16820312 | NA |  0.41661450 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214160 |  0.45258430 | -0.01396400 |  2.95433736 |  0.38630462 | NA |  0.46225259 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214193 | -2.44460797 |  2.13922572 |  0.16586153 | -0.11648754 | NA | -0.71341413 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214194 |          NA | -1.79452336 |  0.32472101 | -2.26251125 | NA | -0.01573116 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214253 | -1.39064407 | -1.29023123 | -1.72721636 | -2.89380598 | NA | -0.40899506 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214274 | -0.31787163 |  1.09867418 |          NA | -0.29028556 | NA | -0.41056791 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214357 | -0.34382582 | -0.17527953 | -0.11419190 |  1.51654112 | NA |  2.03312397 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214413 | -0.81799686 | -1.65288782 | -1.32054901 | -0.96304089 | NA |  0.96759593 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214517 | -0.35869706 |  1.71566033 |  1.27662230 |  1.28102994 | NA |  0.72199219 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214530 |          NA |          NA | -1.90293229 | -0.87361908 | NA |  0.15419285 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214655 |  0.56697595 |  0.91750306 | -0.79633600 |  1.05349636 | NA | -1.15545642 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214706 |  0.57189840 |  1.25597429 | -0.54215997 | -0.58148170 | NA | -0.38550556 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214717 | -1.73147583 | -1.22039437 | -0.68413210 | -0.09883536 | NA | -1.31531465 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214736 | -0.85407978 | -1.68542659 | -0.05669600 | -1.12853837 | NA |  0.15348506 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214753 |  1.03080797 | -0.04829543 | -0.13953565 | -0.38410354 | NA |  0.88414800 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214940 |  0.04631884 | -1.77462924 |  0.76365566 |  0.40670428 | NA |  0.15609336 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214944 |          NA |  0.18850540 |  0.96971118 |  0.28924718 | NA |  0.80388254 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000214967 | -0.31326768 | -1.82583344 |  0.76598984 |  0.21049729 | NA | -0.75453371 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215012 |  0.64287394 |  0.23868078 |  0.21511005 |  0.44112059 | NA |  0.56495368 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215021 |  0.30005670 | -0.16562910 |  0.53408575 |  0.85071629 | NA | -1.27914917 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215041 | -0.99744004 |  0.55672777 |  0.77792305 | -0.40734518 | NA |  2.26693535 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215114 |  0.99766344 | -1.02371836 |  0.99257654 |  0.81619185 | NA |  0.90462214 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215126 |  0.84524620 |  0.56395090 | -0.37646827 |  0.68983483 | NA |  1.48095489 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215193 | -0.06694359 |  0.21530281 | -0.52403498 |  1.69092894 | NA |  1.29820156 | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215251 |          NA |  0.20708491 |  0.25430226 |  0.07474785 | NA | -0.63851118 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215252 |  0.48236033 | -1.26481926 | -0.36134848 |  0.14095147 | NA |  0.71949059 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215301 | -0.97372675 | -0.47196299 | -0.32129517 |  1.17368984 | NA |  0.05126289 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215305 | -0.03369558 | -1.12211978 |  0.50003290 | -0.44566694 | NA | -0.66550738 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| GTEX-1211K | ENSG00000215440 |  0.85139471 |  0.65542102 | -0.12812780 |  0.53822845 | NA |  0.16725840 | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA | NA | NA | 3 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| GTEX-ZVP2 | ENSG00000116031 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000116035 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000116039 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000119457 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000122877 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000124935 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000125571 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000128203 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000130055 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000130475 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000131686 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000131969 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000132677 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000132975 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000133247 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000134765 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000137709 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000138111 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000142046 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000143520 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000145244 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000148291 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000151117 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000154928 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000155367 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000158050 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000158106 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000158220 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE | Inf | 0.991 | test |\n",
       "| GTEX-ZVP2 | ENSG00000158786 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "| GTEX-ZVP2 | ENSG00000159455 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ |    NA |    NA |    NA |    NA |    NA |    NA |    NA |  NA |    NA | test |\n",
       "\n"
      ],
      "text/plain": [
       "        individual gene            zscore.Adipose.20..20Subcutaneous\n",
       "1       GTEX-1211K ENSG00000214113 -0.46848705                      \n",
       "2       GTEX-1211K ENSG00000214114          NA                      \n",
       "3       GTEX-1211K ENSG00000214160  0.45258430                      \n",
       "4       GTEX-1211K ENSG00000214193 -2.44460797                      \n",
       "5       GTEX-1211K ENSG00000214194          NA                      \n",
       "6       GTEX-1211K ENSG00000214253 -1.39064407                      \n",
       "7       GTEX-1211K ENSG00000214274 -0.31787163                      \n",
       "8       GTEX-1211K ENSG00000214357 -0.34382582                      \n",
       "9       GTEX-1211K ENSG00000214413 -0.81799686                      \n",
       "10      GTEX-1211K ENSG00000214517 -0.35869706                      \n",
       "11      GTEX-1211K ENSG00000214530          NA                      \n",
       "12      GTEX-1211K ENSG00000214655  0.56697595                      \n",
       "13      GTEX-1211K ENSG00000214706  0.57189840                      \n",
       "14      GTEX-1211K ENSG00000214717 -1.73147583                      \n",
       "15      GTEX-1211K ENSG00000214736 -0.85407978                      \n",
       "16      GTEX-1211K ENSG00000214753  1.03080797                      \n",
       "17      GTEX-1211K ENSG00000214940  0.04631884                      \n",
       "18      GTEX-1211K ENSG00000214944          NA                      \n",
       "19      GTEX-1211K ENSG00000214967 -0.31326768                      \n",
       "20      GTEX-1211K ENSG00000215012  0.64287394                      \n",
       "21      GTEX-1211K ENSG00000215021  0.30005670                      \n",
       "22      GTEX-1211K ENSG00000215041 -0.99744004                      \n",
       "23      GTEX-1211K ENSG00000215114  0.99766344                      \n",
       "24      GTEX-1211K ENSG00000215126  0.84524620                      \n",
       "25      GTEX-1211K ENSG00000215193 -0.06694359                      \n",
       "26      GTEX-1211K ENSG00000215251          NA                      \n",
       "27      GTEX-1211K ENSG00000215252  0.48236033                      \n",
       "28      GTEX-1211K ENSG00000215301 -0.97372675                      \n",
       "29      GTEX-1211K ENSG00000215305 -0.03369558                      \n",
       "30      GTEX-1211K ENSG00000215440  0.85139471                      \n",
       "⋮       ⋮          ⋮               ⋮                                \n",
       "9203073 GTEX-ZVP2  ENSG00000116031 NA                               \n",
       "9203074 GTEX-ZVP2  ENSG00000116035 NA                               \n",
       "9203075 GTEX-ZVP2  ENSG00000116039 NA                               \n",
       "9203076 GTEX-ZVP2  ENSG00000119457 NA                               \n",
       "9203077 GTEX-ZVP2  ENSG00000122877 NA                               \n",
       "9203078 GTEX-ZVP2  ENSG00000124935 NA                               \n",
       "9203079 GTEX-ZVP2  ENSG00000125571 NA                               \n",
       "9203080 GTEX-ZVP2  ENSG00000128203 NA                               \n",
       "9203081 GTEX-ZVP2  ENSG00000130055 NA                               \n",
       "9203082 GTEX-ZVP2  ENSG00000130475 NA                               \n",
       "9203083 GTEX-ZVP2  ENSG00000131686 NA                               \n",
       "9203084 GTEX-ZVP2  ENSG00000131969 NA                               \n",
       "9203085 GTEX-ZVP2  ENSG00000132677 NA                               \n",
       "9203086 GTEX-ZVP2  ENSG00000132975 NA                               \n",
       "9203087 GTEX-ZVP2  ENSG00000133247 NA                               \n",
       "9203088 GTEX-ZVP2  ENSG00000134765 NA                               \n",
       "9203089 GTEX-ZVP2  ENSG00000137709 NA                               \n",
       "9203090 GTEX-ZVP2  ENSG00000138111 NA                               \n",
       "9203091 GTEX-ZVP2  ENSG00000142046 NA                               \n",
       "9203092 GTEX-ZVP2  ENSG00000143520 NA                               \n",
       "9203093 GTEX-ZVP2  ENSG00000145244 NA                               \n",
       "9203094 GTEX-ZVP2  ENSG00000148291 NA                               \n",
       "9203095 GTEX-ZVP2  ENSG00000151117 NA                               \n",
       "9203096 GTEX-ZVP2  ENSG00000154928 NA                               \n",
       "9203097 GTEX-ZVP2  ENSG00000155367 NA                               \n",
       "9203098 GTEX-ZVP2  ENSG00000158050 NA                               \n",
       "9203099 GTEX-ZVP2  ENSG00000158106 NA                               \n",
       "9203100 GTEX-ZVP2  ENSG00000158220 NA                               \n",
       "9203101 GTEX-ZVP2  ENSG00000158786 NA                               \n",
       "9203102 GTEX-ZVP2  ENSG00000159455 NA                               \n",
       "        zscore.Adipose.20..20Visceral.20.28Omentum.29 zscore.Adrenal.20Gland\n",
       "1       -0.24032781                                   -0.22958386           \n",
       "2        1.23622894                                    0.02992738           \n",
       "3       -0.01396400                                    2.95433736           \n",
       "4        2.13922572                                    0.16586153           \n",
       "5       -1.79452336                                    0.32472101           \n",
       "6       -1.29023123                                   -1.72721636           \n",
       "7        1.09867418                                            NA           \n",
       "8       -0.17527953                                   -0.11419190           \n",
       "9       -1.65288782                                   -1.32054901           \n",
       "10       1.71566033                                    1.27662230           \n",
       "11               NA                                   -1.90293229           \n",
       "12       0.91750306                                   -0.79633600           \n",
       "13       1.25597429                                   -0.54215997           \n",
       "14      -1.22039437                                   -0.68413210           \n",
       "15      -1.68542659                                   -0.05669600           \n",
       "16      -0.04829543                                   -0.13953565           \n",
       "17      -1.77462924                                    0.76365566           \n",
       "18       0.18850540                                    0.96971118           \n",
       "19      -1.82583344                                    0.76598984           \n",
       "20       0.23868078                                    0.21511005           \n",
       "21      -0.16562910                                    0.53408575           \n",
       "22       0.55672777                                    0.77792305           \n",
       "23      -1.02371836                                    0.99257654           \n",
       "24       0.56395090                                   -0.37646827           \n",
       "25       0.21530281                                   -0.52403498           \n",
       "26       0.20708491                                    0.25430226           \n",
       "27      -1.26481926                                   -0.36134848           \n",
       "28      -0.47196299                                   -0.32129517           \n",
       "29      -1.12211978                                    0.50003290           \n",
       "30       0.65542102                                   -0.12812780           \n",
       "⋮       ⋮                                             ⋮                     \n",
       "9203073 NA                                            NA                    \n",
       "9203074 NA                                            NA                    \n",
       "9203075 NA                                            NA                    \n",
       "9203076 NA                                            NA                    \n",
       "9203077 NA                                            NA                    \n",
       "9203078 NA                                            NA                    \n",
       "9203079 NA                                            NA                    \n",
       "9203080 NA                                            NA                    \n",
       "9203081 NA                                            NA                    \n",
       "9203082 NA                                            NA                    \n",
       "9203083 NA                                            NA                    \n",
       "9203084 NA                                            NA                    \n",
       "9203085 NA                                            NA                    \n",
       "9203086 NA                                            NA                    \n",
       "9203087 NA                                            NA                    \n",
       "9203088 NA                                            NA                    \n",
       "9203089 NA                                            NA                    \n",
       "9203090 NA                                            NA                    \n",
       "9203091 NA                                            NA                    \n",
       "9203092 NA                                            NA                    \n",
       "9203093 NA                                            NA                    \n",
       "9203094 NA                                            NA                    \n",
       "9203095 NA                                            NA                    \n",
       "9203096 NA                                            NA                    \n",
       "9203097 NA                                            NA                    \n",
       "9203098 NA                                            NA                    \n",
       "9203099 NA                                            NA                    \n",
       "9203100 NA                                            NA                    \n",
       "9203101 NA                                            NA                    \n",
       "9203102 NA                                            NA                    \n",
       "        zscore.Artery.20..20Aorta zscore.Artery.20..20Coronary\n",
       "1        0.69779491               NA                          \n",
       "2        1.16820312               NA                          \n",
       "3        0.38630462               NA                          \n",
       "4       -0.11648754               NA                          \n",
       "5       -2.26251125               NA                          \n",
       "6       -2.89380598               NA                          \n",
       "7       -0.29028556               NA                          \n",
       "8        1.51654112               NA                          \n",
       "9       -0.96304089               NA                          \n",
       "10       1.28102994               NA                          \n",
       "11      -0.87361908               NA                          \n",
       "12       1.05349636               NA                          \n",
       "13      -0.58148170               NA                          \n",
       "14      -0.09883536               NA                          \n",
       "15      -1.12853837               NA                          \n",
       "16      -0.38410354               NA                          \n",
       "17       0.40670428               NA                          \n",
       "18       0.28924718               NA                          \n",
       "19       0.21049729               NA                          \n",
       "20       0.44112059               NA                          \n",
       "21       0.85071629               NA                          \n",
       "22      -0.40734518               NA                          \n",
       "23       0.81619185               NA                          \n",
       "24       0.68983483               NA                          \n",
       "25       1.69092894               NA                          \n",
       "26       0.07474785               NA                          \n",
       "27       0.14095147               NA                          \n",
       "28       1.17368984               NA                          \n",
       "29      -0.44566694               NA                          \n",
       "30       0.53822845               NA                          \n",
       "⋮       ⋮                         ⋮                           \n",
       "9203073 NA                        NA                          \n",
       "9203074 NA                        NA                          \n",
       "9203075 NA                        NA                          \n",
       "9203076 NA                        NA                          \n",
       "9203077 NA                        NA                          \n",
       "9203078 NA                        NA                          \n",
       "9203079 NA                        NA                          \n",
       "9203080 NA                        NA                          \n",
       "9203081 NA                        NA                          \n",
       "9203082 NA                        NA                          \n",
       "9203083 NA                        NA                          \n",
       "9203084 NA                        NA                          \n",
       "9203085 NA                        NA                          \n",
       "9203086 NA                        NA                          \n",
       "9203087 NA                        NA                          \n",
       "9203088 NA                        NA                          \n",
       "9203089 NA                        NA                          \n",
       "9203090 NA                        NA                          \n",
       "9203091 NA                        NA                          \n",
       "9203092 NA                        NA                          \n",
       "9203093 NA                        NA                          \n",
       "9203094 NA                        NA                          \n",
       "9203095 NA                        NA                          \n",
       "9203096 NA                        NA                          \n",
       "9203097 NA                        NA                          \n",
       "9203098 NA                        NA                          \n",
       "9203099 NA                        NA                          \n",
       "9203100 NA                        NA                          \n",
       "9203101 NA                        NA                          \n",
       "9203102 NA                        NA                          \n",
       "        zscore.Artery.20..20Tibial zscore.Brain.20..20Amygdala\n",
       "1       -1.33789635                NA                         \n",
       "2        0.41661450                NA                         \n",
       "3        0.46225259                NA                         \n",
       "4       -0.71341413                NA                         \n",
       "5       -0.01573116                NA                         \n",
       "6       -0.40899506                NA                         \n",
       "7       -0.41056791                NA                         \n",
       "8        2.03312397                NA                         \n",
       "9        0.96759593                NA                         \n",
       "10       0.72199219                NA                         \n",
       "11       0.15419285                NA                         \n",
       "12      -1.15545642                NA                         \n",
       "13      -0.38550556                NA                         \n",
       "14      -1.31531465                NA                         \n",
       "15       0.15348506                NA                         \n",
       "16       0.88414800                NA                         \n",
       "17       0.15609336                NA                         \n",
       "18       0.80388254                NA                         \n",
       "19      -0.75453371                NA                         \n",
       "20       0.56495368                NA                         \n",
       "21      -1.27914917                NA                         \n",
       "22       2.26693535                NA                         \n",
       "23       0.90462214                NA                         \n",
       "24       1.48095489                NA                         \n",
       "25       1.29820156                NA                         \n",
       "26      -0.63851118                NA                         \n",
       "27       0.71949059                NA                         \n",
       "28       0.05126289                NA                         \n",
       "29      -0.66550738                NA                         \n",
       "30       0.16725840                NA                         \n",
       "⋮       ⋮                          ⋮                          \n",
       "9203073 NA                         NA                         \n",
       "9203074 NA                         NA                         \n",
       "9203075 NA                         NA                         \n",
       "9203076 NA                         NA                         \n",
       "9203077 NA                         NA                         \n",
       "9203078 NA                         NA                         \n",
       "9203079 NA                         NA                         \n",
       "9203080 NA                         NA                         \n",
       "9203081 NA                         NA                         \n",
       "9203082 NA                         NA                         \n",
       "9203083 NA                         NA                         \n",
       "9203084 NA                         NA                         \n",
       "9203085 NA                         NA                         \n",
       "9203086 NA                         NA                         \n",
       "9203087 NA                         NA                         \n",
       "9203088 NA                         NA                         \n",
       "9203089 NA                         NA                         \n",
       "9203090 NA                         NA                         \n",
       "9203091 NA                         NA                         \n",
       "9203092 NA                         NA                         \n",
       "9203093 NA                         NA                         \n",
       "9203094 NA                         NA                         \n",
       "9203095 NA                         NA                         \n",
       "9203096 NA                         NA                         \n",
       "9203097 NA                         NA                         \n",
       "9203098 NA                         NA                         \n",
       "9203099 NA                         NA                         \n",
       "9203100 NA                         NA                         \n",
       "9203101 NA                         NA                         \n",
       "9203102 NA                         NA                         \n",
       "        zscore.Brain.20..20Anterior.20cingulate.20cortex.20.28BA24.29 ⋯\n",
       "1       NA                                                            ⋯\n",
       "2       NA                                                            ⋯\n",
       "3       NA                                                            ⋯\n",
       "4       NA                                                            ⋯\n",
       "5       NA                                                            ⋯\n",
       "6       NA                                                            ⋯\n",
       "7       NA                                                            ⋯\n",
       "8       NA                                                            ⋯\n",
       "9       NA                                                            ⋯\n",
       "10      NA                                                            ⋯\n",
       "11      NA                                                            ⋯\n",
       "12      NA                                                            ⋯\n",
       "13      NA                                                            ⋯\n",
       "14      NA                                                            ⋯\n",
       "15      NA                                                            ⋯\n",
       "16      NA                                                            ⋯\n",
       "17      NA                                                            ⋯\n",
       "18      NA                                                            ⋯\n",
       "19      NA                                                            ⋯\n",
       "20      NA                                                            ⋯\n",
       "21      NA                                                            ⋯\n",
       "22      NA                                                            ⋯\n",
       "23      NA                                                            ⋯\n",
       "24      NA                                                            ⋯\n",
       "25      NA                                                            ⋯\n",
       "26      NA                                                            ⋯\n",
       "27      NA                                                            ⋯\n",
       "28      NA                                                            ⋯\n",
       "29      NA                                                            ⋯\n",
       "30      NA                                                            ⋯\n",
       "⋮       ⋮                                                             ⋱\n",
       "9203073 NA                                                            ⋯\n",
       "9203074 NA                                                            ⋯\n",
       "9203075 NA                                                            ⋯\n",
       "9203076 NA                                                            ⋯\n",
       "9203077 NA                                                            ⋯\n",
       "9203078 NA                                                            ⋯\n",
       "9203079 NA                                                            ⋯\n",
       "9203080 NA                                                            ⋯\n",
       "9203081 NA                                                            ⋯\n",
       "9203082 NA                                                            ⋯\n",
       "9203083 NA                                                            ⋯\n",
       "9203084 NA                                                            ⋯\n",
       "9203085 NA                                                            ⋯\n",
       "9203086 NA                                                            ⋯\n",
       "9203087 NA                                                            ⋯\n",
       "9203088 NA                                                            ⋯\n",
       "9203089 NA                                                            ⋯\n",
       "9203090 NA                                                            ⋯\n",
       "9203091 NA                                                            ⋯\n",
       "9203092 NA                                                            ⋯\n",
       "9203093 NA                                                            ⋯\n",
       "9203094 NA                                                            ⋯\n",
       "9203095 NA                                                            ⋯\n",
       "9203096 NA                                                            ⋯\n",
       "9203097 NA                                                            ⋯\n",
       "9203098 NA                                                            ⋯\n",
       "9203099 NA                                                            ⋯\n",
       "9203100 NA                                                            ⋯\n",
       "9203101 NA                                                            ⋯\n",
       "9203102 NA                                                            ⋯\n",
       "        feature.vep.features.regulatory_region_variant.max\n",
       "1          NA                                             \n",
       "2          NA                                             \n",
       "3       FALSE                                             \n",
       "4          NA                                             \n",
       "5          NA                                             \n",
       "6          NA                                             \n",
       "7          NA                                             \n",
       "8          NA                                             \n",
       "9          NA                                             \n",
       "10      FALSE                                             \n",
       "11      FALSE                                             \n",
       "12         NA                                             \n",
       "13         NA                                             \n",
       "14         NA                                             \n",
       "15         NA                                             \n",
       "16         NA                                             \n",
       "17         NA                                             \n",
       "18      FALSE                                             \n",
       "19         NA                                             \n",
       "20      FALSE                                             \n",
       "21         NA                                             \n",
       "22         NA                                             \n",
       "23      FALSE                                             \n",
       "24         NA                                             \n",
       "25      FALSE                                             \n",
       "26         NA                                             \n",
       "27         NA                                             \n",
       "28         NA                                             \n",
       "29         NA                                             \n",
       "30         NA                                             \n",
       "⋮       ⋮                                                 \n",
       "9203073    NA                                             \n",
       "9203074    NA                                             \n",
       "9203075    NA                                             \n",
       "9203076    NA                                             \n",
       "9203077    NA                                             \n",
       "9203078    NA                                             \n",
       "9203079    NA                                             \n",
       "9203080 FALSE                                             \n",
       "9203081    NA                                             \n",
       "9203082    NA                                             \n",
       "9203083 FALSE                                             \n",
       "9203084    NA                                             \n",
       "9203085    NA                                             \n",
       "9203086 FALSE                                             \n",
       "9203087    NA                                             \n",
       "9203088 FALSE                                             \n",
       "9203089    NA                                             \n",
       "9203090    NA                                             \n",
       "9203091    NA                                             \n",
       "9203092    NA                                             \n",
       "9203093 FALSE                                             \n",
       "9203094    NA                                             \n",
       "9203095    NA                                             \n",
       "9203096 FALSE                                             \n",
       "9203097 FALSE                                             \n",
       "9203098    NA                                             \n",
       "9203099 FALSE                                             \n",
       "9203100 FALSE                                             \n",
       "9203101    NA                                             \n",
       "9203102    NA                                             \n",
       "        feature.vep.features.splice_acceptor_variant.max\n",
       "1          NA                                           \n",
       "2          NA                                           \n",
       "3       FALSE                                           \n",
       "4          NA                                           \n",
       "5          NA                                           \n",
       "6          NA                                           \n",
       "7          NA                                           \n",
       "8          NA                                           \n",
       "9          NA                                           \n",
       "10      FALSE                                           \n",
       "11      FALSE                                           \n",
       "12         NA                                           \n",
       "13         NA                                           \n",
       "14         NA                                           \n",
       "15         NA                                           \n",
       "16         NA                                           \n",
       "17         NA                                           \n",
       "18      FALSE                                           \n",
       "19         NA                                           \n",
       "20      FALSE                                           \n",
       "21         NA                                           \n",
       "22         NA                                           \n",
       "23      FALSE                                           \n",
       "24         NA                                           \n",
       "25      FALSE                                           \n",
       "26         NA                                           \n",
       "27         NA                                           \n",
       "28         NA                                           \n",
       "29         NA                                           \n",
       "30         NA                                           \n",
       "⋮       ⋮                                               \n",
       "9203073    NA                                           \n",
       "9203074    NA                                           \n",
       "9203075    NA                                           \n",
       "9203076    NA                                           \n",
       "9203077    NA                                           \n",
       "9203078    NA                                           \n",
       "9203079    NA                                           \n",
       "9203080 FALSE                                           \n",
       "9203081    NA                                           \n",
       "9203082    NA                                           \n",
       "9203083 FALSE                                           \n",
       "9203084    NA                                           \n",
       "9203085    NA                                           \n",
       "9203086 FALSE                                           \n",
       "9203087    NA                                           \n",
       "9203088 FALSE                                           \n",
       "9203089    NA                                           \n",
       "9203090    NA                                           \n",
       "9203091    NA                                           \n",
       "9203092    NA                                           \n",
       "9203093 FALSE                                           \n",
       "9203094    NA                                           \n",
       "9203095    NA                                           \n",
       "9203096 FALSE                                           \n",
       "9203097 FALSE                                           \n",
       "9203098    NA                                           \n",
       "9203099 FALSE                                           \n",
       "9203100 FALSE                                           \n",
       "9203101    NA                                           \n",
       "9203102    NA                                           \n",
       "        feature.vep.features.splice_donor_variant.max\n",
       "1          NA                                        \n",
       "2          NA                                        \n",
       "3       FALSE                                        \n",
       "4          NA                                        \n",
       "5          NA                                        \n",
       "6          NA                                        \n",
       "7          NA                                        \n",
       "8          NA                                        \n",
       "9          NA                                        \n",
       "10      FALSE                                        \n",
       "11      FALSE                                        \n",
       "12         NA                                        \n",
       "13         NA                                        \n",
       "14         NA                                        \n",
       "15         NA                                        \n",
       "16         NA                                        \n",
       "17         NA                                        \n",
       "18      FALSE                                        \n",
       "19         NA                                        \n",
       "20      FALSE                                        \n",
       "21         NA                                        \n",
       "22         NA                                        \n",
       "23      FALSE                                        \n",
       "24         NA                                        \n",
       "25      FALSE                                        \n",
       "26         NA                                        \n",
       "27         NA                                        \n",
       "28         NA                                        \n",
       "29         NA                                        \n",
       "30         NA                                        \n",
       "⋮       ⋮                                            \n",
       "9203073    NA                                        \n",
       "9203074    NA                                        \n",
       "9203075    NA                                        \n",
       "9203076    NA                                        \n",
       "9203077    NA                                        \n",
       "9203078    NA                                        \n",
       "9203079    NA                                        \n",
       "9203080 FALSE                                        \n",
       "9203081    NA                                        \n",
       "9203082    NA                                        \n",
       "9203083 FALSE                                        \n",
       "9203084    NA                                        \n",
       "9203085    NA                                        \n",
       "9203086 FALSE                                        \n",
       "9203087    NA                                        \n",
       "9203088 FALSE                                        \n",
       "9203089    NA                                        \n",
       "9203090    NA                                        \n",
       "9203091    NA                                        \n",
       "9203092    NA                                        \n",
       "9203093 FALSE                                        \n",
       "9203094    NA                                        \n",
       "9203095    NA                                        \n",
       "9203096 FALSE                                        \n",
       "9203097 FALSE                                        \n",
       "9203098    NA                                        \n",
       "9203099 FALSE                                        \n",
       "9203100 FALSE                                        \n",
       "9203101    NA                                        \n",
       "9203102    NA                                        \n",
       "        feature.vep.features.splice_region_variant.max\n",
       "1          NA                                         \n",
       "2          NA                                         \n",
       "3       FALSE                                         \n",
       "4          NA                                         \n",
       "5          NA                                         \n",
       "6          NA                                         \n",
       "7          NA                                         \n",
       "8          NA                                         \n",
       "9          NA                                         \n",
       "10      FALSE                                         \n",
       "11      FALSE                                         \n",
       "12         NA                                         \n",
       "13         NA                                         \n",
       "14         NA                                         \n",
       "15         NA                                         \n",
       "16         NA                                         \n",
       "17         NA                                         \n",
       "18      FALSE                                         \n",
       "19         NA                                         \n",
       "20      FALSE                                         \n",
       "21         NA                                         \n",
       "22         NA                                         \n",
       "23      FALSE                                         \n",
       "24         NA                                         \n",
       "25      FALSE                                         \n",
       "26         NA                                         \n",
       "27         NA                                         \n",
       "28         NA                                         \n",
       "29         NA                                         \n",
       "30         NA                                         \n",
       "⋮       ⋮                                             \n",
       "9203073    NA                                         \n",
       "9203074    NA                                         \n",
       "9203075    NA                                         \n",
       "9203076    NA                                         \n",
       "9203077    NA                                         \n",
       "9203078    NA                                         \n",
       "9203079    NA                                         \n",
       "9203080 FALSE                                         \n",
       "9203081    NA                                         \n",
       "9203082    NA                                         \n",
       "9203083 FALSE                                         \n",
       "9203084    NA                                         \n",
       "9203085    NA                                         \n",
       "9203086 FALSE                                         \n",
       "9203087    NA                                         \n",
       "9203088 FALSE                                         \n",
       "9203089    NA                                         \n",
       "9203090    NA                                         \n",
       "9203091    NA                                         \n",
       "9203092    NA                                         \n",
       "9203093 FALSE                                         \n",
       "9203094    NA                                         \n",
       "9203095    NA                                         \n",
       "9203096 FALSE                                         \n",
       "9203097 FALSE                                         \n",
       "9203098    NA                                         \n",
       "9203099 FALSE                                         \n",
       "9203100 FALSE                                         \n",
       "9203101    NA                                         \n",
       "9203102    NA                                         \n",
       "        feature.vep.features.stop_gained.max\n",
       "1          NA                               \n",
       "2          NA                               \n",
       "3       FALSE                               \n",
       "4          NA                               \n",
       "5          NA                               \n",
       "6          NA                               \n",
       "7          NA                               \n",
       "8          NA                               \n",
       "9          NA                               \n",
       "10      FALSE                               \n",
       "11      FALSE                               \n",
       "12         NA                               \n",
       "13         NA                               \n",
       "14         NA                               \n",
       "15         NA                               \n",
       "16         NA                               \n",
       "17         NA                               \n",
       "18      FALSE                               \n",
       "19         NA                               \n",
       "20      FALSE                               \n",
       "21         NA                               \n",
       "22         NA                               \n",
       "23      FALSE                               \n",
       "24         NA                               \n",
       "25      FALSE                               \n",
       "26         NA                               \n",
       "27         NA                               \n",
       "28         NA                               \n",
       "29         NA                               \n",
       "30         NA                               \n",
       "⋮       ⋮                                   \n",
       "9203073    NA                               \n",
       "9203074    NA                               \n",
       "9203075    NA                               \n",
       "9203076    NA                               \n",
       "9203077    NA                               \n",
       "9203078    NA                               \n",
       "9203079    NA                               \n",
       "9203080 FALSE                               \n",
       "9203081    NA                               \n",
       "9203082    NA                               \n",
       "9203083 FALSE                               \n",
       "9203084    NA                               \n",
       "9203085    NA                               \n",
       "9203086 FALSE                               \n",
       "9203087    NA                               \n",
       "9203088 FALSE                               \n",
       "9203089    NA                               \n",
       "9203090    NA                               \n",
       "9203091    NA                               \n",
       "9203092    NA                               \n",
       "9203093 FALSE                               \n",
       "9203094    NA                               \n",
       "9203095    NA                               \n",
       "9203096 FALSE                               \n",
       "9203097 FALSE                               \n",
       "9203098    NA                               \n",
       "9203099 FALSE                               \n",
       "9203100 FALSE                               \n",
       "9203101    NA                               \n",
       "9203102    NA                               \n",
       "        feature.vep.features.synonymous_variant.max\n",
       "1          NA                                      \n",
       "2          NA                                      \n",
       "3       FALSE                                      \n",
       "4          NA                                      \n",
       "5          NA                                      \n",
       "6          NA                                      \n",
       "7          NA                                      \n",
       "8          NA                                      \n",
       "9          NA                                      \n",
       "10      FALSE                                      \n",
       "11      FALSE                                      \n",
       "12         NA                                      \n",
       "13         NA                                      \n",
       "14         NA                                      \n",
       "15         NA                                      \n",
       "16         NA                                      \n",
       "17         NA                                      \n",
       "18      FALSE                                      \n",
       "19         NA                                      \n",
       "20       TRUE                                      \n",
       "21         NA                                      \n",
       "22         NA                                      \n",
       "23      FALSE                                      \n",
       "24         NA                                      \n",
       "25      FALSE                                      \n",
       "26         NA                                      \n",
       "27         NA                                      \n",
       "28         NA                                      \n",
       "29         NA                                      \n",
       "30         NA                                      \n",
       "⋮       ⋮                                          \n",
       "9203073    NA                                      \n",
       "9203074    NA                                      \n",
       "9203075    NA                                      \n",
       "9203076    NA                                      \n",
       "9203077    NA                                      \n",
       "9203078    NA                                      \n",
       "9203079    NA                                      \n",
       "9203080 FALSE                                      \n",
       "9203081    NA                                      \n",
       "9203082    NA                                      \n",
       "9203083 FALSE                                      \n",
       "9203084    NA                                      \n",
       "9203085    NA                                      \n",
       "9203086 FALSE                                      \n",
       "9203087    NA                                      \n",
       "9203088 FALSE                                      \n",
       "9203089    NA                                      \n",
       "9203090    NA                                      \n",
       "9203091    NA                                      \n",
       "9203092    NA                                      \n",
       "9203093 FALSE                                      \n",
       "9203094    NA                                      \n",
       "9203095    NA                                      \n",
       "9203096 FALSE                                      \n",
       "9203097 FALSE                                      \n",
       "9203098    NA                                      \n",
       "9203099 FALSE                                      \n",
       "9203100 FALSE                                      \n",
       "9203101    NA                                      \n",
       "9203102    NA                                      \n",
       "        feature.vep.features.upstream_gene_variant.max\n",
       "1          NA                                         \n",
       "2          NA                                         \n",
       "3       FALSE                                         \n",
       "4          NA                                         \n",
       "5          NA                                         \n",
       "6          NA                                         \n",
       "7          NA                                         \n",
       "8          NA                                         \n",
       "9          NA                                         \n",
       "10      FALSE                                         \n",
       "11      FALSE                                         \n",
       "12         NA                                         \n",
       "13         NA                                         \n",
       "14         NA                                         \n",
       "15         NA                                         \n",
       "16         NA                                         \n",
       "17         NA                                         \n",
       "18      FALSE                                         \n",
       "19         NA                                         \n",
       "20      FALSE                                         \n",
       "21         NA                                         \n",
       "22         NA                                         \n",
       "23      FALSE                                         \n",
       "24         NA                                         \n",
       "25      FALSE                                         \n",
       "26         NA                                         \n",
       "27         NA                                         \n",
       "28         NA                                         \n",
       "29         NA                                         \n",
       "30         NA                                         \n",
       "⋮       ⋮                                             \n",
       "9203073    NA                                         \n",
       "9203074    NA                                         \n",
       "9203075    NA                                         \n",
       "9203076    NA                                         \n",
       "9203077    NA                                         \n",
       "9203078    NA                                         \n",
       "9203079    NA                                         \n",
       "9203080 FALSE                                         \n",
       "9203081    NA                                         \n",
       "9203082    NA                                         \n",
       "9203083 FALSE                                         \n",
       "9203084    NA                                         \n",
       "9203085    NA                                         \n",
       "9203086 FALSE                                         \n",
       "9203087    NA                                         \n",
       "9203088 FALSE                                         \n",
       "9203089    NA                                         \n",
       "9203090    NA                                         \n",
       "9203091    NA                                         \n",
       "9203092    NA                                         \n",
       "9203093 FALSE                                         \n",
       "9203094    NA                                         \n",
       "9203095    NA                                         \n",
       "9203096  TRUE                                         \n",
       "9203097  TRUE                                         \n",
       "9203098    NA                                         \n",
       "9203099  TRUE                                         \n",
       "9203100  TRUE                                         \n",
       "9203101    NA                                         \n",
       "9203102    NA                                         \n",
       "        feature.vep.features.sift_score.pval_max_significant\n",
       "1       NA                                                  \n",
       "2       NA                                                  \n",
       "3       NA                                                  \n",
       "4       NA                                                  \n",
       "5       NA                                                  \n",
       "6       NA                                                  \n",
       "7       NA                                                  \n",
       "8       NA                                                  \n",
       "9       NA                                                  \n",
       "10      NA                                                  \n",
       "11      NA                                                  \n",
       "12      NA                                                  \n",
       "13      NA                                                  \n",
       "14      NA                                                  \n",
       "15      NA                                                  \n",
       "16      NA                                                  \n",
       "17      NA                                                  \n",
       "18      NA                                                  \n",
       "19      NA                                                  \n",
       "20      NA                                                  \n",
       "21      NA                                                  \n",
       "22      NA                                                  \n",
       "23      NA                                                  \n",
       "24      NA                                                  \n",
       "25      NA                                                  \n",
       "26      NA                                                  \n",
       "27      NA                                                  \n",
       "28      NA                                                  \n",
       "29      NA                                                  \n",
       "30      NA                                                  \n",
       "⋮       ⋮                                                   \n",
       "9203073  NA                                                 \n",
       "9203074  NA                                                 \n",
       "9203075  NA                                                 \n",
       "9203076  NA                                                 \n",
       "9203077  NA                                                 \n",
       "9203078  NA                                                 \n",
       "9203079  NA                                                 \n",
       "9203080  NA                                                 \n",
       "9203081  NA                                                 \n",
       "9203082  NA                                                 \n",
       "9203083  NA                                                 \n",
       "9203084  NA                                                 \n",
       "9203085  NA                                                 \n",
       "9203086  NA                                                 \n",
       "9203087  NA                                                 \n",
       "9203088  NA                                                 \n",
       "9203089  NA                                                 \n",
       "9203090  NA                                                 \n",
       "9203091  NA                                                 \n",
       "9203092  NA                                                 \n",
       "9203093  NA                                                 \n",
       "9203094  NA                                                 \n",
       "9203095  NA                                                 \n",
       "9203096  NA                                                 \n",
       "9203097  NA                                                 \n",
       "9203098  NA                                                 \n",
       "9203099  NA                                                 \n",
       "9203100 Inf                                                 \n",
       "9203101  NA                                                 \n",
       "9203102  NA                                                 \n",
       "        feature.vep.features.polyphen_score.max fold\n",
       "1       NA                                      3   \n",
       "2       NA                                      3   \n",
       "3       NA                                      3   \n",
       "4       NA                                      3   \n",
       "5       NA                                      3   \n",
       "6       NA                                      3   \n",
       "7       NA                                      3   \n",
       "8       NA                                      3   \n",
       "9       NA                                      3   \n",
       "10      NA                                      3   \n",
       "11      NA                                      3   \n",
       "12      NA                                      3   \n",
       "13      NA                                      3   \n",
       "14      NA                                      3   \n",
       "15      NA                                      3   \n",
       "16      NA                                      3   \n",
       "17      NA                                      3   \n",
       "18      NA                                      3   \n",
       "19      NA                                      3   \n",
       "20      NA                                      3   \n",
       "21      NA                                      3   \n",
       "22      NA                                      3   \n",
       "23      NA                                      3   \n",
       "24      NA                                      3   \n",
       "25      NA                                      3   \n",
       "26      NA                                      3   \n",
       "27      NA                                      3   \n",
       "28      NA                                      3   \n",
       "29      NA                                      3   \n",
       "30      NA                                      3   \n",
       "⋮       ⋮                                       ⋮   \n",
       "9203073    NA                                   test\n",
       "9203074    NA                                   test\n",
       "9203075    NA                                   test\n",
       "9203076    NA                                   test\n",
       "9203077    NA                                   test\n",
       "9203078    NA                                   test\n",
       "9203079    NA                                   test\n",
       "9203080    NA                                   test\n",
       "9203081    NA                                   test\n",
       "9203082    NA                                   test\n",
       "9203083    NA                                   test\n",
       "9203084    NA                                   test\n",
       "9203085    NA                                   test\n",
       "9203086    NA                                   test\n",
       "9203087    NA                                   test\n",
       "9203088    NA                                   test\n",
       "9203089    NA                                   test\n",
       "9203090    NA                                   test\n",
       "9203091    NA                                   test\n",
       "9203092    NA                                   test\n",
       "9203093    NA                                   test\n",
       "9203094    NA                                   test\n",
       "9203095    NA                                   test\n",
       "9203096    NA                                   test\n",
       "9203097    NA                                   test\n",
       "9203098    NA                                   test\n",
       "9203099    NA                                   test\n",
       "9203100 0.991                                   test\n",
       "9203101    NA                                   test\n",
       "9203102    NA                                   test"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_dt = (\n",
    "    open_dataset(snakemake@input$subset_pq)\n",
    "    %>% select(`individual`, `gene`)\n",
    "    %>% unique()\n",
    "    %>% collect()\n",
    "    %>% left_join(\n",
    "        open_dataset(snakemake@input$data_pq) %>% collect(),\n",
    "        by=c(\"individual\", \"gene\")\n",
    "    )\n",
    "    # %>% select(`subtissue`, `tissue`, `hilo_padj`)\n",
    "    # %>% rename(tissue_type=tissue, tissue=subtissue)\n",
    "    %>% as.data.table()\n",
    ")\n",
    "data_dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a0899769",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove infinite values\n",
    "data_dt$`feature.vep.features.sift_score.pval_max_significant`[is.infinite(data_dt$`feature.vep.features.sift_score.pval_max_significant`)] = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e06eef7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>&lt;NA&gt;</li><li>FALSE</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item <NA>\n",
       "\\item FALSE\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. &lt;NA&gt;\n",
       "2. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]    NA FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_dt[, unique(`feature.vep.features.TF_binding_site_variant.max`)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c60b8c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>&lt;NA&gt;</li><li>FALSE</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item <NA>\n",
       "\\item FALSE\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. &lt;NA&gt;\n",
       "2. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]    NA FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_dt[, unique(`feature.vep.features.regulatory_region_variant.max`)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4df6a802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>&lt;NA&gt;</li><li>FALSE</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item <NA>\n",
       "\\item FALSE\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. &lt;NA&gt;\n",
       "2. FALSE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]    NA FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_dt[, unique(`feature.vep.features.intergenic_variant.max`)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6816191d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dt = data_dt[, !c(\n",
    "    \"feature.vep.features.TF_binding_site_variant.max\",\n",
    "    \"feature.vep.features.regulatory_region_variant.max\",\n",
    "    \"feature.vep.features.intergenic_variant.max\"\n",
    ")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b72cbdfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Command line arguments\n",
    "#########################################\n",
    "# arguments <- parse_args(OptionParser(usage = \"%prog [options]\", description=\"Watershed command line args\",\n",
    "# \toption_list=list(\n",
    "# \t\tmake_option(c(\"-t\",\"--training_input\"), default = NULL, help=\"The Watershed input file containing instances used to train the model [default %default]\"),\n",
    "# # \t\tmake_option(c(\"-i\",\"--prediction_input\"), default = NULL, help=\"The Watershed input file containing instances to predict on\"),\n",
    "# \t\tmake_option(c(\"-d\",\"--number_dimensions\"), default=1, help=\"The number of outlier variables.\"),\n",
    "# \t\tmake_option(c(\"-m\",\"--model_name\"), default=\"Watershed_exact\", help=\"Name of model. Options are Watershed_exact, Watershed_approximate, and RIVER\"),\n",
    "# \t\tmake_option(c(\"-p\",\"--dirichlet_prior_parameter\"), default=10, help=\"Parameter defining Dirichlet distribution the acts as a prior a Phi (the model parameters defining E|Z\"),\n",
    "# \t\tmake_option(c(\"-l\",\"--l2_prior_parameter\"), default=.01, help=\"Parameter defining L2 (gaussian) distribution the acts as a prior on the parameters of the conditional random Field (the model parameters defining Z|G\"),\n",
    "# \t\tmake_option(c(\"-o\",\"--output_file\"), default=\"watershed_prediction_object.rds\", help=\"file name to save results to\"),\n",
    "# \t\tmake_option(c(\"-b\",\"--binary_pvalue_threshold\"), default=.1, help=\"Absolute p-value threshold used to create binary outliers used for Genomic Annotation Model\")\n",
    "# \t)\n",
    "# ))\n",
    "# process command line args\n",
    "# training_input_file <- arguments$training_input\n",
    "# prediction_input_file <- arguments$prediction_input\n",
    "# number_of_dimensions <- arguments$number_dimensions\n",
    "model_name <- \"Watershed_approximate\" # arguments$model_name\n",
    "pseudoc <- 10 # arguments$dirichlet_prior_parameter\n",
    "lambda_init <- .01 # arguments$l2_prior_parameter\n",
    "if (lambda_init == \"NA\") {\n",
    "\tlambda_init <- NA\n",
    "}\n",
    "# # output_stem <- arguments$output_prefix\n",
    "# binary_pvalue_threshold <- arguments$binary_pvalue_threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ab66c91b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################\n",
    "# Fixed Parameters\n",
    "#########################################\n",
    "# If arguments$l2_prior_parameter == NA, perform grid search over the following values of lambda to determine optimal lambda\n",
    "lambda_costs <- c(.1, .01, 1e-3)\n",
    "# If arguments$l2_prior_parameter == NA, Number of folds to be used in K-fold cross validation for Genomic annotation model ()\n",
    "nfolds <- 5\n",
    "# Parameters used for Variational Optimization (only applies if arguments$model_name==\"Watershed_approximate\")\n",
    "vi_step_size <- .8\n",
    "vi_threshold <- 1e-8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8b4a6b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set seed for reproducability (feel free to change!)\n",
    "set.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0fbc3581",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Load in and parse Watershed input file\n",
    "load_watershed_data <- function(data_dt, use_adjusted_pvalues=FALSE, pvalue_threshold=0.05, num_dims=NULL, num_features=NULL) {\n",
    "    target_dt = data_dt %>% select(all_of(colnames(data_dt)[startsWith(colnames(data_dt), \"y_truecat\")]))\n",
    "    pval_dt = data_dt %>% select(all_of(colnames(data_dt)[startsWith(colnames(data_dt), \"pval\")]))\n",
    "    padj_dt = data_dt %>% select(all_of(colnames(data_dt)[startsWith(colnames(data_dt), \"padj\")]))\n",
    "    feat_dt = data_dt %>% select(all_of(colnames(data_dt)[startsWith(colnames(data_dt), \"feature\")]))\n",
    "    \n",
    "    if (is.null(num_dims)) {\n",
    "        number_of_dimensions = length(colnames(pval_dt))\n",
    "    } else{\n",
    "        number_of_dimensions = num_dims\n",
    "        target_dt = target_dt[, 1:num_dims]\n",
    "        pval_dt = pval_dt[, 1:num_dims]\n",
    "        padj_dt = padj_dt[, 1:num_dims]\n",
    "    }\n",
    "    if (! is.null(num_features)) {\n",
    "        feat_dt = feat_dt[, 1:num_features]\n",
    "    }\n",
    "\n",
    "    rownames = data_dt[,paste(`individual`, \":\", `gene`, sep=\"\")]\n",
    "\n",
    "    feat = as.matrix(feat_dt)\n",
    "    feat[is.na(feat)] = 0\n",
    "    feat[is.nan(feat)] = 0\n",
    "    rownames(feat) <- rownames\n",
    "\n",
    "    if (use_adjusted_pvalues) {\n",
    "        print(\"Using adjusted p-values...\")\n",
    "        # use adjusted p-values\n",
    "        outlier_pvalues = as.matrix(padj_dt)\n",
    "    } else {\n",
    "        # use nominal p-values\n",
    "        print(\"Using nominal p-values...\")\n",
    "        outlier_pvalues = as.matrix(pval_dt)\n",
    "    }\n",
    "    rownames(outlier_pvalues) <- rownames\n",
    "\n",
    "    outliers_binary <- ifelse(abs(outlier_pvalues) <= pvalue_threshold, 1, 0)\n",
    "    # Convert outlier status into discretized random variables\n",
    "    outliers_discrete <- get_discretized_outliers(outlier_pvalues)\n",
    "    # Put all data into compact data structure\n",
    "    data_input <- list(\n",
    "        feat=as.matrix(feat),\n",
    "        outliers_binary=as.matrix(outliers_binary),\n",
    "        outliers_discrete=outliers_discrete,\n",
    "        number_of_dimensions=number_of_dimensions\n",
    "    )\n",
    "    return(data_input)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b3a4c50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "learn_watershed_model_parameters_from_training_data <- function(data_input, model_name, pseudoc, lambda_init, binary_pvalue_threshold, lambda_costs, nfolds, vi_step_size, vi_threshold) {\n",
    "\t# Parse data_input for relevent data\n",
    "\tfeat_all <- data_input$feat\n",
    "\tdiscrete_outliers_all <- data_input$outliers_discrete\n",
    "\tbinary_outliers_all <- data_input$outliers_binary\n",
    "    number_of_dimensions <- data_input$number_of_dimensions\n",
    "\n",
    "\t#######################################\n",
    "\t## Standardize Genomic Annotations (features)\n",
    "\t#######################################\n",
    "\tmean_feat <- apply(feat_all, 2, mean)\n",
    "\tsd_feat <- apply(feat_all, 2, sd)\n",
    " \tfeat_all <- scale(feat_all, center=mean_feat, scale=sd_feat)\n",
    "\n",
    "  \t#######################################\n",
    "\t## Fit Genomic Annotation Model (GAM)\n",
    "\t#######################################\n",
    "    # print(feat_all)\n",
    "\tgam_data <- logistic_regression_genomic_annotation_model_cv(feat_all, binary_outliers_all, nfolds, lambda_costs, lambda_init)\n",
    "\t# Report optimal lambda learned from cross-validation data (if applicable)\n",
    "\tif (is.na(lambda_init)) {\n",
    "\t\tcat(paste0(nfolds,\"-fold cross validation on GAM yielded optimal lambda of \", gam_data$lambda, \"\\n\"))\n",
    "\t}\n",
    "\n",
    "\t#######################################\n",
    "\t### Initialize phi using GAM\n",
    "\t#######################################\n",
    "\t# Compute GAM Predictions on data via function in  CPP file (\"independent_crf_exact_updates.cpp\")\n",
    "\tgam_posterior_obj <- update_independent_marginal_probabilities_exact_inference_cpp(\n",
    "        feat_all,\n",
    "        binary_outliers_all,\n",
    "        gam_data$gam_parameters$theta_singleton,\n",
    "        gam_data$gam_parameters$theta_pair,\n",
    "        gam_data$gam_parameters$theta,\n",
    "        matrix(0,2,2),\n",
    "        matrix(0,2,2),\n",
    "        number_of_dimensions,\n",
    "        choose(number_of_dimensions, 2),\n",
    "        FALSE\n",
    "    )\n",
    "\tgam_posteriors <- gam_posterior_obj$probability\n",
    "\t# Initialize Phi using GAM posteriors\n",
    "\t# ie. Compute MAP estimates of the coefficients defined by P(outlier_status| FR)\n",
    "\tphi_init <- map_phi_initialization(discrete_outliers_all, gam_posteriors, number_of_dimensions, pseudoc)\n",
    "\n",
    "\t#######################################\n",
    "\t### Fit Watershed Model\n",
    "\t#######################################\n",
    "\twatershed_model <- train_watershed_model(\n",
    "        feat_all,\n",
    "        discrete_outliers_all,\n",
    "        phi_init,\n",
    "        gam_data$gam_parameters$theta_pair,\n",
    "        gam_data$gam_parameters$theta_singleton,\n",
    "        gam_data$gam_parameters$theta,\n",
    "        pseudoc,\n",
    "        gam_data$lambda,\n",
    "        number_of_dimensions,\n",
    "        model_name,\n",
    "        vi_step_size,\n",
    "        vi_threshold\n",
    "    )\n",
    "\n",
    "\treturn(list(\n",
    "        mean_feat=mean_feat,\n",
    "        sd_feat=sd_feat,\n",
    "        model_params=watershed_model,\n",
    "        gam_model_params=gam_data\n",
    "    ))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c0aa0eef",
   "metadata": {},
   "outputs": [],
   "source": [
    "predict_watershed = function(watershed_object, prediction_data_input) {\n",
    "    # Parse data_input for relevent data\n",
    "    predictions_feat <- prediction_data_input$feat\n",
    "    # predictions_discretized_outliers <- prediction_data_input$outliers_discrete\n",
    "    # Scale prediction features (according to mean and standard deviation from training data)\n",
    "    predictions_feat <- scale(predictions_feat, center=watershed_object$mean_feat, scale=watershed_object$sd_feat)\n",
    "    number_dimensions <- prediction_data_input$number_of_dimensions\n",
    "\n",
    "\n",
    "    ########################\n",
    "    ## Inference to compute Watershed posterior probabilities\n",
    "    ########################\n",
    "    # watershed_info <- update_marginal_posterior_probabilities(predictions_feat, predictions_discretized_outliers, watershed_object$model_params)\n",
    "    # watershed_posteriors <- watershed_info$probability  # Marginal posteriors\n",
    "\n",
    "    # We do not need Expression for inference, therefore replace `predictions_discretized_outliers` with an empty `matrix()`.\n",
    "    watershed_z_given_g = update_conditional_z_given_g_probabilities(predictions_feat, matrix(), watershed_object$model_params)\n",
    "    watershed_z_given_g_posteriors = watershed_z_given_g$probability\n",
    "    colnames(watershed_z_given_g_posteriors) = watershed_object$pred_names\n",
    "\n",
    "    # ########################\n",
    "    # # Add row names and column names to posterior predictions matrix\n",
    "    # ########################\n",
    "    # watershed_z_given_g_posteriors_mat <- cbind(rownames(predictions_feat), watershed_z_given_g_posteriors)\n",
    "    # colnames(watershed_z_given_g_posteriors_mat) = c(\"sample_names\", paste0(\"watershed_z_given_g_posterior_outlier_signal_\", 1:number_dimensions))\n",
    "    \n",
    "    return(watershed_z_given_g_posteriors)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "09b0170a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# watershed_data = load_watershed_data(\n",
    "#     data_dt,\n",
    "#     use_adjusted_pvalues=TRUE\n",
    "#     pvalue_threshold=0.2\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1981f132",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(paste(snakemake@params$output_basedir, \"fold\", sep=\"/\"), showWarnings=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "17b158fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'3'</li><li>'2'</li><li>'0'</li><li>'test'</li><li>'1'</li><li>'4'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '3'\n",
       "\\item '2'\n",
       "\\item '0'\n",
       "\\item 'test'\n",
       "\\item '1'\n",
       "\\item '4'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '3'\n",
       "2. '2'\n",
       "3. '0'\n",
       "4. 'test'\n",
       "5. '1'\n",
       "6. '4'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"3\"    \"2\"    \"0\"    \"test\" \"1\"    \"4\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_dt[, unique(`fold`)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9a413efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_any <- function(dt, cols) {\n",
    "  return(dt[, Reduce(`|`, .SD), .SDcols=cols])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1042f7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_all_na <- function(dt, cols) {\n",
    "  return(dt[, Reduce(`&`, lapply(.SD, is.na)), .SDcols=cols])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f64002be",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold = function(target_fold, subsample_training_negative_class=NULL) {\n",
    "    pvalue_threshold=0.2\n",
    "    use_adjusted_pvalues = FALSE\n",
    "    # # limit dimensions and features for testing\n",
    "    # num_dims=3\n",
    "    # num_features=3\n",
    "    num_dims=NULL\n",
    "    num_features=NULL\n",
    "    \n",
    "    print(paste0(\"Reading fold \", target_fold, \"...\"))\n",
    "\n",
    "    training_data = data_dt %>% filter(`fold` != .env$target_fold)\n",
    "    if (! is.null(subsample_training_negative_class)) {\n",
    "        print(paste0(\"Subsampling non-outliers of training data by '\", subsample_training_negative_class, \"x' the number of outliers...\"))\n",
    "        training_data$has_outlier=row_any(training_data, cols=colnames(training_data)[startsWith(colnames(training_data), \"y_truecat\")])\n",
    "        num_has_outlier = length(which(training_data$has_outlier))\n",
    "        num_sample = min(subsample_training_negative_class * num_has_outlier, length(training_data$has_outlier) - num_has_outlier)\n",
    "        \n",
    "        training_data = rbind(\n",
    "            training_data[(`has_outlier` != TRUE) | is.na(`has_outlier`)][sample(.N, num_sample)],\n",
    "            training_data[`has_outlier` == TRUE]\n",
    "        )[order(`individual`, `gene`)]\n",
    "    }\n",
    "    training_watershed_data = load_watershed_data(\n",
    "        training_data,\n",
    "        use_adjusted_pvalues=use_adjusted_pvalues,\n",
    "        pvalue_threshold=pvalue_threshold,\n",
    "        num_features=num_features,\n",
    "        num_dims=num_dims\n",
    "    )\n",
    "\n",
    "    testing_data = data_dt %>% filter(`fold` == .env$target_fold)\n",
    "    testing_watershed_data = load_watershed_data(\n",
    "        testing_data,\n",
    "        use_adjusted_pvalues=use_adjusted_pvalues,\n",
    "        pvalue_threshold=pvalue_threshold,\n",
    "        num_features=num_features,\n",
    "        num_dims=num_dims\n",
    "    )\n",
    "\n",
    "    #######################################\n",
    "    ## Train Watershed model on training data\n",
    "    #######################################\n",
    "    print(\"Training watershed...\")\n",
    "    watershed_object <- learn_watershed_model_parameters_from_training_data(training_watershed_data, model_name, pseudoc, lambda_init, binary_pvalue_threshold, lambda_costs, nfolds, vi_step_size, vi_threshold)\n",
    "    watershed_object$pred_names = colnames(training_watershed_data$outliers_binary)\n",
    "\n",
    "    #######################################\n",
    "    ## Calculate predictions\n",
    "    #######################################\n",
    "    train_predictions = cbind(\n",
    "        training_data %>% select(`individual`, `gene`),\n",
    "        as.data.table(predict_watershed(watershed_object, training_watershed_data))\n",
    "    )\n",
    "    # train_predictions\n",
    "\n",
    "    test_predictions = cbind(\n",
    "        testing_data %>% select(`individual`, `gene`),\n",
    "        as.data.table(predict_watershed(watershed_object, testing_watershed_data))\n",
    "    )\n",
    "    # test_predictions\n",
    "\n",
    "    #######################################\n",
    "    ## Save predictions\n",
    "    #######################################\n",
    "    print(paste0(\"Writing to fold \", target_fold, \"...\"))\n",
    "    \n",
    "    fold_dir = paste(snakemake@params$output_basedir, \"fold\", target_fold, sep=\"/\")\n",
    "    dir.create(fold_dir, showWarnings=FALSE)\n",
    "    \n",
    "    saveRDS(watershed_object, paste(fold_dir, \"model.rds\", sep=\"/\"))\n",
    "    write_parquet(train_predictions, paste(fold_dir, \"train_predictions.parquet\", sep=\"/\"))\n",
    "    write_parquet(test_predictions, paste(fold_dir, \"test_predictions.parquet\", sep=\"/\"))\n",
    "    \n",
    "    return(list(\n",
    "        training_data=training_data,\n",
    "        testing_data=testing_data,\n",
    "        watershed_object=watershed_object,\n",
    "        training_predictions=train_predictions,\n",
    "        testing_predictions=test_predictions\n",
    "    ))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77af2eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "subsample_training_negative_class = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fd743597",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'3'</li><li>'2'</li><li>'0'</li><li>'test'</li><li>'1'</li><li>'4'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '3'\n",
       "\\item '2'\n",
       "\\item '0'\n",
       "\\item 'test'\n",
       "\\item '1'\n",
       "\\item '4'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '3'\n",
       "2. '2'\n",
       "3. '0'\n",
       "4. 'test'\n",
       "5. '1'\n",
       "6. '4'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"3\"    \"2\"    \"0\"    \"test\" \"1\"    \"4\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "folds = data_dt[[\"fold\"]] %>% unique()\n",
    "folds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "846ad408",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('0', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8edd04",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('1', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dccb47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('2', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba57bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('3', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92b49443",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('4', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97cc4c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_fold('test', subsample_training_negative_class=subsample_training_negative_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa46496f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "formats": "ipynb,auto:percent",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R [conda env:anaconda-ukbb-trait-analysis-R]",
   "language": "R",
   "name": "conda-env-anaconda-ukbb-trait-analysis-R-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
